File: test/fixtures/swift/complexity.swift
// Fixture for the Swift structural rules. Every callable here has a
// hand-counted cyclomatic complexity asserted in test/rules.test.ts.
//
// Keep this file boring: any construct added here changes the expected counts.

// CC 1 — no decision points at all.
func plain() -> Int {
	return 1
}

// CC 11. Decision points, in order:
//   1 guard, 2 if, 3 &&, 4 else-if, 5 for, 6 where, 7 nested if,
//   8 case, 9 case, 10 ternary
// `default:` is deliberately present and deliberately NOT counted.
// At 11 this is just over the Eq. 3 cutoff, so it MUST count as eroded — the
// mirror of the TypeScript fixture's boundary case at exactly 10.
func branching(_ n: Int) -> String {
	guard n >= 0 else { return "neg" }

	if n > 0 && n < 10 {
		return "small"
	} else if n >= 10 {
		return "large"
	}

	for x in [1, 2] where x > 0 {
		if x == 1 { return "one" }
	}

	switch n {
	case 1: return "a"
	case 2: return "b"
	default: break
	}

	return n > 0 ? "t" : "f"
}

// CC 3 — one `??` and one catch block.
func risky(_ value: String?) -> String {
	do {
		return try compute(value) ?? "none"
	} catch {
		return "err"
	}
}

// CC 2, and exactly ONE callable. The trailing closure is not a callable, so
// its ternary folds into this function. This is the case that matters most in
// SwiftUI code, where closures are everywhere.
func withClosure(_ xs: [Int]) -> [Int] {
	return xs.map { $0 > 0 ? $0 : -$0 }
}

struct Badge {
	let label: String

	// CC 2 — a computed property is a callable, so SwiftUI `body` complexity is
	// measured rather than invisible.
	var body: String {
		return label.isEmpty ? "empty" : label
	}

	// CC 2 — init_declaration with one guard-free branch.
	init(label: String) {
		if label.isEmpty {
			self.label = "none"
		} else {
			self.label = label
		}
	}
}

func compute(_ value: String?) throws -> String? {
	return value
}

File: test/fixtures/typescript/complexity.ts
// Fixture for the TypeScript structural rules. Every callable here has a
// hand-counted cyclomatic complexity asserted in test/rules.test.ts.
//
// Keep this file boring: any construct added here changes the expected counts.

// CC 1 — no decision points at all.
export function plain(): number {
	return 1;
}

// CC 10. Decision points, in order:
//   1 if, 2 &&, 3 else-if, 4 for, 5 nested if, 6 case, 7 case, 8 ternary, 9 catch
// `default:` is deliberately present and deliberately NOT counted.
// Sitting exactly at 10 also pins the Eq. 3 boundary: >10 is high-complexity,
// so this callable must NOT be counted as eroded.
export function branching(n: number): string {
	if (n > 0 && n < 10) {
		return "small";
	} else if (n >= 10) {
		return "large";
	}

	for (const x of [1, 2]) {
		if (x) return "loop";
	}

	switch (n) {
		case 1:
			return "one";
		case 2:
			return "two";
		default:
			break;
	}

	try {
		return n ? "t" : "f";
	} catch {
		return "err";
	}
}

// CC 2, and exactly ONE callable. The two inline closures are not callables in
// their own right, so the ternary inside the first one folds into this function
// rather than becoming a separate near-zero-complexity entry.
export function withClosure(items: number[]): number[] {
	return items.filter((x) => (x > 0 ? true : false)).map((x) => x * 2);
}

// TWO callables. `inner` is bound to a name, so it gets its own entry (CC 2 for
// the ternary) and its complexity is removed from `outer` (CC 2 for the if) —
// the nested-function rule Radon applies.
export function outer(n: number): number {
	const inner = (x: number): number => (x > 0 ? x : -x);
	if (n > 0) return inner(n);
	return 0;
}

export class Widget {
	// CC 2 — a method_definition, one ternary.
	render(flag: boolean): string {
		return flag ? "on" : "off";
	}

	// CC 1 — `constructor` is also a method_definition.
	constructor(readonly id: string) {}
}

// CC 3 — `??` and `||` both count, `?.` deliberately does not.
export function nullish(a: string | null, b: string | null): string {
	return a?.trim() ?? b ?? "fallback";
}
